Source/WebCore/Modules/mediastream/gstreamer/GStreamerIceAgent.cpp

```diff
@@ -0,0 +1,783 @@
+        if (key == "transport"_s) {
+            transport = value;
+            break;
+        }
+    }
+    if (!transport || transport == "udp"_s)
+        return "udp"_s;
+    if (!transport || transport == "tcp"_s)
+        return "tcp"_s;
+
+    return "none"_s;
+}
+
+static gboolean webkitGstWebRTCIceAgentGetSelectedPair(GstWebRTCICE* ice, GstWebRTCICEStream* stream, GstWebRTCICECandidateStats** localStats, GstWebRTCICECandidateStats** remoteStats)
+{
+    if (!stream)
+        return FALSE;
+
+    auto result = webkitGstWebRTCIceStreamGetSelectedPair(WEBKIT_GST_WEBRTC_ICE_STREAM(stream), localStats, remoteStats);
+    if (!result)
+        return result;
+
+    auto backend = WEBKIT_GST_WEBRTC_ICE_BACKEND(ice);
+    auto relayProtocol = getRelayProtocol(backend);
+    (*localStats)->relay_proto = relayProtocol;
+    (*remoteStats)->relay_proto = relayProtocol;
+    return TRUE;
+}
+
+void webkitGstWebRTCIceAgentClosed(WebKitGstIceAgent* agent)
+{
+    agent->priv->agentIsClosed.exchange(true);
+    agent->priv->streams.clear();
+
+    if (!agent->priv->closePromise)
+        return;
+
+    gst_promise_reply(agent->priv->closePromise.get(), nullptr);
+    agent->priv->closePromise.clear();
+}
+
+#if GST_CHECK_VERSION(1, 27, 0)
+static void webkitGstWebRTCIceAgentClose(GstWebRTCICE* ice, GstPromise* promise)
+{
+    auto backend = WEBKIT_GST_WEBRTC_ICE_BACKEND(ice);
+
+    auto isClosed = backend->priv->agentIsClosed.load();
+    if (isClosed)
+        return;
+
+    bool shouldWait = promise == nullptr;
+    backend->priv->closePromise = adoptGRef(promise);
+    auto now = WTF::MonotonicTime::now().secondsSinceEpoch();
+    rice_agent_close(backend->priv->agent.get(), now.nanoseconds());
+    webkitGstWebRTCIceAgentWakeup(backend);
+
+    isClosed = backend->priv->agentIsClosed.load();
+    if (!shouldWait || isClosed)
+        return;
+
+    while (!isClosed) {
+        g_main_context_iteration(backend->priv->runLoop->mainContext(), FALSE);
+        isClosed = backend->priv->agentIsClosed.load();
+    }
+}
+#endif
+
+static void webkitGstWebRTCIceAgentFinalize(GObject* object)
+{
+    auto backend = WEBKIT_GST_WEBRTC_ICE_BACKEND(object);
+
+    if (backend->priv->onCandidateNotify)
+        backend->priv->onCandidateNotify(backend->priv->onCandidateData);
+    G_OBJECT_CLASS(webkit_gst_webrtc_ice_backend_parent_class)->finalize(object);
+}
+
+static void webkitGstWebRTCIceAgentConstructed(GObject* object)
+{
+    G_OBJECT_CLASS(webkit_gst_webrtc_ice_backend_parent_class)->constructed(object);
+
+    auto backend = WEBKIT_GST_WEBRTC_ICE_BACKEND(object);
+    auto priv = backend->priv;
+
+    priv->agentIsClosed.exchange(false);
+
+    static Atomic<uint32_t> counter = 0;
+    auto id = counter.load();
+    auto threadName = makeString("webrtc-rice-"_s, id);
+    counter.exchangeAdd(1);
+
+    priv->runLoop = RunLoop::create(ASCIILiteral::fromLiteralUnsafe(threadName.ascii().data()));
```

Comment:
You can't do this because of lifetimes. ASCIILiterals are supposed to live forever as the underlying literal is also supposed to live forever. In this case you are creating something that is supposed to live forever from something that is going to be destroyed when the line finishes execution.

I don't know what the proper solution for this is but definitely creating a literal from something it is not a literal literal is not right.